src/kubernetes.go

```diff
@@ -0,0 +1,107 @@
+package src
+
+import (
+	"bufio"
+	"context"
+	"github.com/vshn/provider-exoscale/apis"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/rest"
+	"net/url"
+	"os"
+	ctrl "sigs.k8s.io/controller-runtime"
+	kubernetes "sigs.k8s.io/controller-runtime/pkg/client"
+	"strings"
+)
+
+type Cluster struct {
+	url   url.URL
+	token string
+}
+
+type K8sClient struct {
+	kubernetes.Client
+	ClusterHostname string
+}
+
+// ParseClusterConfigFile accepts a file with a list of cluster to be parsed
+// Each cluster entry should have a pair of <server>=<token> per line
+func ParseClusterConfigFile(ctx context.Context, clusterFilePath string) (*[]Cluster, error) {
+	log := ctrl.LoggerFrom(ctx)
+	var clusters []Cluster
+	file, err := os.Open(clusterFilePath)
+	if err != nil {
+		log.V(1).Error(err, "Cannot open file with cluster info")
+		return nil, err
+	}
+	defer closeFile(ctx, file)
+	scanner := bufio.NewScanner(file)
+	for scanner.Scan() {
+		line := scanner.Text()
+		clusters = append(clusters, *parseLine(ctx, line))
+	}
+	if err = scanner.Err(); err != nil {
+		log.V(1).Error(err, "Cannot read cluster info file")
+		return nil, err
+	}
+	return &clusters, err
+}
+
+// CreateKubernetesClients creates as many k8s clients as there are cluster entries from the parsed file
+// The function skips clients that cannot be initialized
+func CreateKubernetesClients(ctx context.Context, clusters *[]Cluster) (*[]K8sClient, error) {
+	log := ctrl.LoggerFrom(ctx)
+	var k8sClients []K8sClient
+
+	scheme := runtime.NewScheme()
+	err := apis.AddToScheme(scheme)
+	if err != nil {
+		log.V(1).Error(err, "Cannot add k8s exoscale to scheme")
+		return nil, err
+	}
+	for _, cluster := range *clusters {
+		config := rest.Config{Host: cluster.url.String(), BearerToken: cluster.token}
+		client, err := kubernetes.New(&config, kubernetes.Options{
+			Scheme: scheme,
+		})
+		if err != nil {
+			// Continue in case a k8s client cannot be initiated.
+			log.V(1).Error(err, "Cannot initiate k8s client, skipping config", "cluster", cluster.url)
+			continue
+		}
+		k8sClients = append(k8sClients, K8sClient{client, cluster.url.Hostname()})
+	}
+	return &k8sClients, nil
+}
+
+func parseLine(ctx context.Context, line string) *Cluster {
+	log := ctrl.LoggerFrom(ctx)
+	if line == "" {
+		log.V(2).Info("Empty line")
+		return nil
+	}
+	k8sConfig := strings.Split(line, "=")
+	if len(k8sConfig) < 2 {
+		log.V(2).Info("Invalid line", line)
+		return nil
+	}
+	clusterUrl, err := url.Parse(k8sConfig[0])
+	if err != nil {
+		log.V(2).Info("Invalid cluster url", line)
+		return nil
+	}
+	clusterToken := k8sConfig[1]
+	return &Cluster{
+		url:   *clusterUrl,
+		token: clusterToken,
+	}
+}
```

Comment:
As mentioned earlier, please use json or (preferably yaml parser).
I can already figure out a number of issues of this parser:
- The URL can contain an `=` sign, so splitting by `=` leads to errors.
- The token could contain an `=` (or multiple) as well, so you'd get multiple "lines".
- No unit tests

Reply:
Yep that was easier then expected, moved to yaml file.